packages/core/src/render3/node_manipulation.ts

```diff
@@ -452,27 +448,31 @@ function processCleanups(tView: TView, lView: LView): void {
         ngDevMode && assertNumber(targetIdx, 'cleanup target must be a number');
         if (targetIdx >= 0) {
           // unregister
-          lCleanup[lastLCleanupIndex = targetIdx]();
+          lCleanup[targetIdx]();
         } else {
           // Subscription
-          lCleanup[lastLCleanupIndex = -targetIdx].unsubscribe();
+          lCleanup[-targetIdx].unsubscribe();
         }
         i += 2;
       } else {
         // This is a cleanup function that is grouped with the index of its context
-        const context = lCleanup[lastLCleanupIndex = tCleanup[i + 1]];
+        const context = lCleanup[tCleanup[i + 1]];
         tCleanup[i].call(context);
       }
     }
   }
   if (lCleanup !== null) {
-    for (let i = lastLCleanupIndex + 1; i < lCleanup.length; i++) {
-      const instanceCleanupFn = lCleanup[i];
-      ngDevMode && assertFunction(instanceCleanupFn, 'Expecting instance cleanup function.');
-      instanceCleanupFn();
-    }
     lView[CLEANUP] = null;
   }
+  const destroyHooks = lView[ON_DESTROY_HOOKS];
+  if (destroyHooks !== null) {
+    for (let i = 0; i < destroyHooks.length; i++) {
+      const destroyHooksFn = destroyHooks[i];
+      ngDevMode && assertFunction(destroyHooksFn, 'Expecting destroy hook to be a function.');
+      destroyHooksFn();
+    }
```

Comment:
```suggestion
    for (const destroyHookFn of destroyHooks) {
      ngDevMode && assertFunction(destroyHooksFn, 'Expecting destroy hook to be a function.');
      destroyHooksFn();
    }
```
Why not use a `for of` loop?

Reply:
Mostly to avoid the iterator protocol as it is reasonably "hot" code path and I didn't feel like it negatively impacts readability here. Would prefer to leave it as-is if you don't mind.